In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import json
import string
import re
import pandas as pd
import numpy as np
from collections import Counter

In [3]:
import nltk
from nltk.sentiment import SentimentIntensityAnalyzer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('vader_lexicon')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


True

In [4]:
def read_json_to_dataframe(file_path):
    try:
        df = pd.read_json(file_path, lines=True)
        return df

    except Exception as e:
        print(f"Error reading file: {e}")
        return None

def remove_apostrophes(series):
    return series.apply(lambda lst: [re.sub(r"'+", '', word) \
                                         if word.count("'") >= 2 else word for word in lst])

def De_symbolize_and_split(df, column_name, new_column_name, separator):
    df[new_column_name] = (df[column_name].str.lower()
                           .str.replace(r"[^a-zA-Z' ]", ' ', regex=True)
                           .str.replace(r'\s+', ' ', regex=True)
                           .str.strip()
                           .str.split(separator))
    return df

def sum_votes(vote_dict):
    weights = {'funny': 1, 'useful': 2, 'cool': 1}
    return sum(vote_dict.get(key, 0) * weight for key, weight in weights.items())

sia = SentimentIntensityAnalyzer()
stop_words = set(stopwords.words('english'))

def evaluative_words(words):
    return [word for word in words if word not in stop_words \
            and sia.polarity_scores(word)['compound'] != 0]

In [5]:
df_business = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_training_set_business.json")
df_checkin = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_training_set_checkin.json")
df_review = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_training_set_review.json")
df_user = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_training_set_user.json")

In [6]:
df_review = De_symbolize_and_split(df_review, 'text', 'split_text',' ')
df_review['split_text'] = remove_apostrophes(df_review['split_text'])
df_review['votes_weight'] = df_review['votes'].apply(sum_votes)
df_review['text_length'] = df_review['text'].apply(len)

In [7]:
df_review['evaluative_words'] = df_review['split_text'].apply(evaluative_words)
df_words_counts = df_review["evaluative_words"].explode().value_counts()
top_words = set(df_words_counts.head(100).index.tolist())
df_review['top_words_count'] = df_review['evaluative_words'] \
    .apply(lambda words: sum(word in top_words for word in words) if isinstance(words, list) else 0)

In [8]:
df_review

votes                 user_id  \
0       {'funny': 0, 'useful': 5, 'cool': 2}  rLtl8ZkDX5vH5nAx9C3q5Q   
1       {'funny': 0, 'useful': 0, 'cool': 0}  0a2KyEL0d3Yb1V6aivbIuQ   
2       {'funny': 0, 'useful': 1, 'cool': 0}  0hT2KtfLiobPvh6cDC8JQg   
3       {'funny': 0, 'useful': 2, 'cool': 1}  uZetl9T0NcROGOyFfughhg   
4       {'funny': 0, 'useful': 0, 'cool': 0}  vYmM4KTsC8ZfQBg-j5MWkw   
...                                      ...                     ...   
229902  {'funny': 0, 'useful': 0, 'cool': 0}  6e7pZofhDuIlD_rX2oYirQ   
229903  {'funny': 0, 'useful': 2, 'cool': 0}  dDNfSFT0VApxPmURclX6_g   
229904  {'funny': 0, 'useful': 0, 'cool': 0}  M5wHt6Odh1k5v0tIjqd8DQ   
229905  {'funny': 1, 'useful': 2, 'cool': 0}  jopndPrv-H5KW2CfScnw9A   
229906  {'funny': 1, 'useful': 0, 'cool': 1}  mbRUG6h0Mgb2XIJvscIoMg   

                     review_id  stars       date  \
0       fWKvX83p0-ka4JS3dc6E5A      5 2011-01-26   
1       IjZ33sJrzXqU-0X6U8NwyA      5 2011-07-27   
2       IESLBzqUCLdSzSqm0eCSxQ      4 2012-06-14   
3       G-WvGaISbqqaMHlNnByodA      5 2010-05-27   
4       1uJFq2r5QfJG_6ExMRCaGw      5 2012-01-05   
...                        ...    ...        ...   
229902  f9JaiNg_FMoPNWxt7MlbZQ      2 2011-04-14   
229903  QDWRP1pW5r0huIBAoGmFyg      1 2011-01-23   
229904  JmR3yk7JlS1LVVxtIc3xBQ      4 2010-10-11   
229905  z5b2p5TbCg0uaIiIe8n62w      3 2011-01-18   
229906  QM1rFJsW-ZJoCHbgsysKaw      5 2010-08-14   

                                                     text    type  \
0       My wife took me here on my birthday for breakf...  review   
1       I have no idea why some people give bad review...  review   
2       love the gyro plate. Rice is so good and I als...  review   
3       Rosie, Dakota, and I LOVE Chaparral Dog Park!!...  review   
4       General Manager Scott Petello is a good egg!!!...  review   
...                                                   ...     ...   
229902  I really wanted to like this place because it'...  review   
229903  My husband I stayed here for two nights.  Of c...  review   
229904  Cool atmosphere. A lot of beers on tap and goo...  review   
229905  I have to take a star off for the spotty servi...  review   
229906                                       So cool, yo.  review   

                   business_id  \
0       9yKzy9PApeiPPOUJEtnvkg   
1       ZRJwVLyzEJq1VAihDhYiow   
2       6oRAC4uyJCsJl1X0WZpVSA   
3       _1QQZuf4zZOyFCvXc0o6Vg   
4       6ozycU1RpktNG2-1BroVtw   
...                        ...   
229902  vnffHkFJbmd-J3OaBbK2Eg   
229903  l5oUrgQ190l8CcN8uzd_pA   
229904  -EctXOb3B7T177jGYUhjVA   
229905  YQvg0JCGRFUkb6reMMf3Iw   
229906  gKxOZvQTTd5hpFa3r5igGQ   

                                               split_text  votes_weight  \
0       [my, wife, took, me, here, on, my, birthday, f...            12   
1       [i, have, no, idea, why, some, people, give, b...             0   
2       [love, the, gyro, plate, rice, is, so, good, a...             2   
3       [rosie, dakota, and, i, love, chaparral, dog, ...             5   
4       [general, manager, scott, petello, is, a, good...             0   
...                                                   ...           ...   
229902  [i, really, wanted, to, like, this, place, bec...             0   
229903  [my, husband, i, stayed, here, for, two, night...             4   
229904  [cool, atmosphere, a, lot, of, beers, on, tap,...             0   
229905  [i, have, to, take, a, star, off, for, the, sp...             5   
229906                                     [so, cool, yo]             2   

        text_length                                   evaluative_words  \
0               889  [excellent, perfect, pleasure, excellent, like...   
1              1345  [bad, please, fault, like, friend, pretty, ple...   
2                76                                       [love, good]   
3               419               [love, wonderful, clean, huge, play]   
4               469  [good, assure, trea

In [9]:
#Convert to numeric representation
df_business['isOpen'] = df_business['open'].astype(int)

#Count the total number of checkin days
df_checkin['checkin_nums'] = df_checkin['checkin_info'].apply(lambda x: sum(x.values()))

In [10]:
# Calculate the total votes
df_user['votes_total'] = df_user['votes'].apply(lambda x: sum(x.values()))

# Calculate the ratio of votes to review_count
df_user['votes_per_review'] = df_user['votes_total'] / df_user['review_count']

# If division by 0
df_user['votes_per_review'] = df_user['votes_per_review'].replace([pd.NaT, pd.NaT], 0)

In [11]:
df_review

votes                 user_id  \
0       {'funny': 0, 'useful': 5, 'cool': 2}  rLtl8ZkDX5vH5nAx9C3q5Q   
1       {'funny': 0, 'useful': 0, 'cool': 0}  0a2KyEL0d3Yb1V6aivbIuQ   
2       {'funny': 0, 'useful': 1, 'cool': 0}  0hT2KtfLiobPvh6cDC8JQg   
3       {'funny': 0, 'useful': 2, 'cool': 1}  uZetl9T0NcROGOyFfughhg   
4       {'funny': 0, 'useful': 0, 'cool': 0}  vYmM4KTsC8ZfQBg-j5MWkw   
...                                      ...                     ...   
229902  {'funny': 0, 'useful': 0, 'cool': 0}  6e7pZofhDuIlD_rX2oYirQ   
229903  {'funny': 0, 'useful': 2, 'cool': 0}  dDNfSFT0VApxPmURclX6_g   
229904  {'funny': 0, 'useful': 0, 'cool': 0}  M5wHt6Odh1k5v0tIjqd8DQ   
229905  {'funny': 1, 'useful': 2, 'cool': 0}  jopndPrv-H5KW2CfScnw9A   
229906  {'funny': 1, 'useful': 0, 'cool': 1}  mbRUG6h0Mgb2XIJvscIoMg   

                     review_id  stars       date  \
0       fWKvX83p0-ka4JS3dc6E5A      5 2011-01-26   
1       IjZ33sJrzXqU-0X6U8NwyA      5 2011-07-27   
2       IESLBzqUCLdSzSqm0eCSxQ      4 2012-06-14   
3       G-WvGaISbqqaMHlNnByodA      5 2010-05-27   
4       1uJFq2r5QfJG_6ExMRCaGw      5 2012-01-05   
...                        ...    ...        ...   
229902  f9JaiNg_FMoPNWxt7MlbZQ      2 2011-04-14   
229903  QDWRP1pW5r0huIBAoGmFyg      1 2011-01-23   
229904  JmR3yk7JlS1LVVxtIc3xBQ      4 2010-10-11   
229905  z5b2p5TbCg0uaIiIe8n62w      3 2011-01-18   
229906  QM1rFJsW-ZJoCHbgsysKaw      5 2010-08-14   

                                                     text    type  \
0       My wife took me here on my birthday for breakf...  review   
1       I have no idea why some people give bad review...  review   
2       love the gyro plate. Rice is so good and I als...  review   
3       Rosie, Dakota, and I LOVE Chaparral Dog Park!!...  review   
4       General Manager Scott Petello is a good egg!!!...  review   
...                                                   ...     ...   
229902  I really wanted to like this place because it'...  review   
229903  My husband I stayed here for two nights.  Of c...  review   
229904  Cool atmosphere. A lot of beers on tap and goo...  review   
229905  I have to take a star off for the spotty servi...  review   
229906                                       So cool, yo.  review   

                   business_id  \
0       9yKzy9PApeiPPOUJEtnvkg   
1       ZRJwVLyzEJq1VAihDhYiow   
2       6oRAC4uyJCsJl1X0WZpVSA   
3       _1QQZuf4zZOyFCvXc0o6Vg   
4       6ozycU1RpktNG2-1BroVtw   
...                        ...   
229902  vnffHkFJbmd-J3OaBbK2Eg   
229903  l5oUrgQ190l8CcN8uzd_pA   
229904  -EctXOb3B7T177jGYUhjVA   
229905  YQvg0JCGRFUkb6reMMf3Iw   
229906  gKxOZvQTTd5hpFa3r5igGQ   

                                               split_text  votes_weight  \
0       [my, wife, took, me, here, on, my, birthday, f...            12   
1       [i, have, no, idea, why, some, people, give, b...             0   
2       [love, the, gyro, plate, rice, is, so, good, a...             2   
3       [rosie, dakota, and, i, love, chaparral, dog, ...             5   
4       [general, manager, scott, petello, is, a, good...             0   
...                                                   ...           ...   
229902  [i, really, wanted, to, like, this, place, bec...             0   
229903  [my, husband, i, stayed, here, for, two, night...             4   
229904  [cool, atmosphere, a, lot, of, beers, on, tap,...             0   
229905  [i, have, to, take, a, star, off, for, the, sp...             5   
229906                                     [so, cool, yo]             2   

        text_length                                   evaluative_words  \
0               889  [excellent, perfect, pleasure, excellent, like...   
1              1345  [bad, please, fault, like, friend, pretty, ple...   
2                76                                       [love, good]   
3               419               [love, wonderful, clean, huge, play]   
4               469  [good, assure, trea

In [12]:
temp_review = df_review[['user_id', 'business_id', 'stars', 'votes_weight', 'text_length', 'top_words_count']].copy()
temp_business = df_business[['business_id', 'review_count', 'isOpen']].copy()
temp_checkin = df_checkin[['business_id', 'checkin_nums']].copy()
temp_user = df_user[['user_id', 'votes_per_review']].copy()

merged_df = pd.merge(temp_review, temp_business, on = 'business_id', how = 'inner')
merged_df = pd.merge(merged_df, temp_user, on = 'user_id', how = 'inner')
merged_df = pd.merge(merged_df, temp_checkin, on = 'business_id', how = 'inner')

In [37]:
merged_df

NameError: ignored

In [4]:
import pandas as pd

In [42]:
merged_df.to_csv('/content/drive/MyDrive/ee562/merged_df_yuhua_train.csv', index=False)

NameError: ignored

In [43]:
merged_df_train = pd.read_csv('/content/drive/MyDrive/ee562/merged_df_yuhua_train.csv')

In [44]:
# split the data
from sklearn.model_selection import train_test_split

# Assuming your DataFrame is named 'df'
train_data, test_data = train_test_split(merged_df_train, test_size=0.2, random_state=42)

In [45]:
# select features and target
# Features (X)
features = train_data[['stars', 'text_length', 'top_words_count', 'review_count', 'isOpen', 'checkin_nums']]
# Target (y)
target = train_data['votes_weight']

#here is the training session for different models
##Model 1: xgboost

In [46]:
!pip install xgboost


In [47]:
from sklearn.model_selection import train_test_split

X = merged_df_train[['stars', 'text_length', 'top_words_count', 'review_count', 'isOpen', 'votes_per_review', 'checkin_nums']]  # Features
y = merged_df_train['votes_weight']  # Target variable


# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [48]:
import xgboost as xgb

# Create XGBoost DMatrix
dtrain = xgb.DMatrix(X_train, label=y_train)

# Set XGBoost parameters
params = {
    'objective': 'reg:squarederror',  # For regression tasks
    'eval_metric': 'rmsle',  # Root Mean Squared Logarithmic Error
    # Add other parameters as needed
}

# Train the XGBoost model
model = xgb.train(params, dtrain, num_boost_round=100)

In [49]:
# Create XGBoost DMatrix for test set
dtest = xgb.DMatrix(X_test)

# Make predictions
predictions = model.predict(dtest)

In [50]:
from sklearn.metrics import mean_squared_error

# Calculate RMSE (Root Mean Squared Error)
rmse = mean_squared_error(y_test, predictions, squared=False)
print(f"Root Mean Squared Error: {rmse}")


Root Mean Squared Error: 5.25815316469066


##The result of XGboost Model is accurate, the error is 5.26, a good result for regression problem in traditional machine learning models

## Model 2: Neural Network Model

In [51]:
# try the nerual network method
!pip install tensorflow

In [53]:
import tensorflow as tf
from tensorflow import keras
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler

X1 = merged_df_train[['stars', 'text_length', 'top_words_count', 'review_count', 'isOpen',  'checkin_nums']]  # Features
y1 = merged_df_train['votes_weight']  # Target variable


# Split the data into training and testing sets
X_train1, X_test1, y_train1, y_test1 = train_test_split(X1, y1, test_size=0.2, random_state=42)


# Standardize the data (optional but recommended for neural networks)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train1)
X_test_scaled = scaler.transform(X_test1)

# Build the neural network model
model = keras.Sequential([
    keras.layers.Dense(64, activation='relu', input_shape=(X_train1.shape[1],)),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dense(1)  # Output layer for regression task
])

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Train the neural network model
model.fit(X_train_scaled, y_train1, epochs=10, batch_size=32, validation_split=0.2)

# Evaluate the model on the test set
predictions = model.predict(X_test_scaled)
rmsle = mean_squared_error(y_test1, predictions, squared=False)
print(f"Root Mean Squared Error: {rmsle}")

Epoch 1/10
4010/4010 [==============================] - 46s 11ms/step - loss: 54.8332 - val_loss: 55.4160
Epoch 2/10
4010/4010 [==============================] - 9s 2ms/step - loss: 54.2165 - val_loss: 55.2648
Epoch 3/10
4010/4010 [==============================] - 12s 3ms/step - loss: 54.1433 - val_loss: 55.4577
Epoch 4/10
4010/4010 [==============================] - 13s 3ms/step - loss: 54.1426 - val_loss: 55.9044
Epoch 5/10
4010/4010 [==============================] - 9s 2ms/step - loss: 54.0977 - val_loss: 55.6234
Epoch 6/10
4010/4010 [==============================] - 11s 3ms/step - loss: 54.1026 - val_loss: 55.2862
Epoch 7/10
4010/4010 [==============================] - 8s 2ms/step - loss: 54.0918 - val_loss: 55.2323
Epoch 8/10
4010/4010 [==============================] - 11s 3ms/step - loss: 54.0693 - val_loss: 55.2383
Epoch 9/10
4010/4010 [==============================] - 11s 3ms/step - loss: 54.0594 - val_loss: 55.1874
Epoch 10/10
1253/1253 [==============================] - 

## result for nn model is slightly better than xgboost
The neural network has an input layer with as many neurons as there are features in the input data.
It contains one hidden layer with 64 neurons and another hidden layer with 32 neurons, both using the ReLU activation function.
The output layer has a single neuron for regression output, and it does not use any activation function.

# Model 3
## knn model

In [54]:
from sklearn.neighbors import KNeighborsRegressor
knn_model = KNeighborsRegressor(n_neighbors=5)

In [55]:
knn_model.fit(X_train_scaled, y_train1)

KNeighborsRegressor()

In [56]:
predictions = knn_model.predict(X_test_scaled)

In [57]:
# Evaluate the model
mse = mean_squared_error(y_test1, predictions)
rmse = mean_squared_error(y_test1, predictions, squared=False)

In [58]:
print(f"Mean Squared Error: {mse}")
print(f"Root Mean Squared Error: {rmse}")

Mean Squared Error: 63.84254994388328
Root Mean Squared Error: 7.990153311663255


## Knn model is slightly worse than the former models

# Test the models on test set

## first load the test set

In [59]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [60]:
import json
import string
import re
import pandas as pd
import numpy as np
from collections import Counter

In [61]:
import nltk
from nltk.sentiment import SentimentIntensityAnalyzer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('vader_lexicon')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [62]:
def read_json_to_dataframe(file_path):
    try:
        df = pd.read_json(file_path, lines=True)
        return df

    except Exception as e:
        print(f"Error reading file: {e}")
        return None

def remove_apostrophes(series):
    return series.apply(lambda lst: [re.sub(r"'+", '', word) \
                                         if word.count("'") >= 2 else word for word in lst])

def De_symbolize_and_split(df, column_name, new_column_name, separator):
    df[new_column_name] = (df[column_name].str.lower()
                           .str.replace(r"[^a-zA-Z' ]", ' ', regex=True)
                           .str.replace(r'\s+', ' ', regex=True)
                           .str.strip()
                           .str.split(separator))
    return df

def sum_votes(vote_dict):
    weights = {'funny': 1, 'useful': 2, 'cool': 1}
    return sum(vote_dict.get(key, 0) * weight for key, weight in weights.items())

sia = SentimentIntensityAnalyzer()
stop_words = set(stopwords.words('english'))

def evaluative_words(words):
    return [word for word in words if word not in stop_words \
            and sia.polarity_scores(word)['compound'] != 0]

In [63]:
df_business = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_test_set_business.json")
df_checkin = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_test_set_checkin.json")
df_review = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_test_set_review.json")
df_user = read_json_to_dataframe("/content/drive/MyDrive/ee562/yelp_test_set_user.json")

In [64]:
df_review

user_id               review_id  \
0      2WkM3pYfx7bt46tv7u4hHA  1Lq3ghX6yzJ7MsWXDkgIGQ   
1      eHWbF0k5QOBLgQXhGdeHmg  zulCqAo_XY9wAefJ58H9Fw   
2      HrjjHfDGTafXyKpQKNrYHg  ptm1X6ReMEYg1Y203KXkxQ   
3      DrWLhrK8WMZf7Jb-Oqc7ww  JCCe3m0LK7dGGxO4ntoGLg   
4      jDCONTPR6nyc3J7iimwzkQ  CeOpuUly_s75QSuHKtfxng   
...                       ...                     ...   
22951  UaIq_zJxnVxU_HfvC16i_A  xxt23yRZ-h95mNiu5SqNpA   
22952  XA_lIArLNis1tzrqSywzHQ  iuv2QctuvIPIXJwZoOgNMg   
22953  TVL1e0NHhxAE6PpeXGfTOA  CT9navwsc9EP2HeimvRpUg   
22954  v7pgiW6IS8jSkQ2B4jeXxg  6en_lbX5obNzwIo06PnarA   
22955  ZIcb9cCwcV4My87A-tPM7Q  EWmbTlSEOEStMVGtKnUJbQ   

                                                    text  \
0      Nice place, big patio.  Now offering LIVE sket...   
1      Friendly staff. Make sure you order the gyro p...   
2      LOVE LOVE LOVE this place for breakfast. They ...   
3      Disgusting sandwich. I should have known bette...   
4      Always a fan of Cafe Zupas and their very frie...   
...                                                  ...   
22951  For a wine bar I would have expected more by-t...   
22952  This place has AWESOME kimchi Chigae, kalbi is...   
22953  LOVE this place!  I recently did a search for ...   
22954  Excellent!  Inexpensive and hot fresh food. Wo...   
22955  We come here often as our favorite dish to eat...   

                  business_id  stars       date    type  
0      AuMz7XGkjLcIUurp_AD51w      5 2010-11-15  review  
1      8i5hB_dmf33NVbWE5SwoMQ      5 2012-09-07  review  
2      nvaAUTTl7oqiJDhuimNG6A      5 2012-12-29  review  
3      QwaoxP5Mgm3PJuZo_4bFsw      1 2013-02-22  review  
4      0lEp4vISRmOXa8Xz2pWhbw      4 2013-02-20  review  
...                       ...    ...        ...     ...  
22951  s1dex3Z3QoqiK7V-zXUgAw      4 2013-02-19  review  
22952  PX3e3qtBx_5VC3vqq47jpQ      4 2012-09-24  review  
22953  SyXz4OwSNxfbszTHP-d2aA      5 2013-02-05  review  
22954  soH8ekCER45AjfMaM6cJ2A      4 2011-06-28  review  
22955  lHr-aMBHh29zLpuB2ShAMQ      3 2012-08-24  review  

[22956 rows x 7 columns]

In [65]:
df_review = De_symbolize_and_split(df_review, 'text', 'split_text',' ')
df_review['split_text'] = remove_apostrophes(df_review['split_text'])
df_review['text_length'] = df_review['text'].apply(len)

In [66]:
df_review['evaluative_words'] = df_review['split_text'].apply(evaluative_words)
df_words_counts = df_review["evaluative_words"].explode().value_counts()
top_words = set(df_words_counts.head(100).index.tolist())
df_review['top_words_count'] = df_review['evaluative_words'] \
    .apply(lambda words: sum(word in top_words for word in words) if isinstance(words, list) else 0)

In [67]:
#Convert to numeric representation
df_business['isOpen'] = df_business['open'].astype(int)

#Count the total number of checkin days
df_checkin['checkin_nums'] = df_checkin['checkin_info'].apply(lambda x: sum(x.values()))

In [68]:
df_review

user_id               review_id  \
0      2WkM3pYfx7bt46tv7u4hHA  1Lq3ghX6yzJ7MsWXDkgIGQ   
1      eHWbF0k5QOBLgQXhGdeHmg  zulCqAo_XY9wAefJ58H9Fw   
2      HrjjHfDGTafXyKpQKNrYHg  ptm1X6ReMEYg1Y203KXkxQ   
3      DrWLhrK8WMZf7Jb-Oqc7ww  JCCe3m0LK7dGGxO4ntoGLg   
4      jDCONTPR6nyc3J7iimwzkQ  CeOpuUly_s75QSuHKtfxng   
...                       ...                     ...   
22951  UaIq_zJxnVxU_HfvC16i_A  xxt23yRZ-h95mNiu5SqNpA   
22952  XA_lIArLNis1tzrqSywzHQ  iuv2QctuvIPIXJwZoOgNMg   
22953  TVL1e0NHhxAE6PpeXGfTOA  CT9navwsc9EP2HeimvRpUg   
22954  v7pgiW6IS8jSkQ2B4jeXxg  6en_lbX5obNzwIo06PnarA   
22955  ZIcb9cCwcV4My87A-tPM7Q  EWmbTlSEOEStMVGtKnUJbQ   

                                                    text  \
0      Nice place, big patio.  Now offering LIVE sket...   
1      Friendly staff. Make sure you order the gyro p...   
2      LOVE LOVE LOVE this place for breakfast. They ...   
3      Disgusting sandwich. I should have known bette...   
4      Always a fan of Cafe Zupas and their very frie...   
...                                                  ...   
22951  For a wine bar I would have expected more by-t...   
22952  This place has AWESOME kimchi Chigae, kalbi is...   
22953  LOVE this place!  I recently did a search for ...   
22954  Excellent!  Inexpensive and hot fresh food. Wo...   
22955  We come here often as our favorite dish to eat...   

                  business_id  stars       date    type  \
0      AuMz7XGkjLcIUurp_AD51w      5 2010-11-15  review   
1      8i5hB_dmf33NVbWE5SwoMQ      5 2012-09-07  review   
2      nvaAUTTl7oqiJDhuimNG6A      5 2012-12-29  review   
3      QwaoxP5Mgm3PJuZo_4bFsw      1 2013-02-22  review   
4      0lEp4vISRmOXa8Xz2pWhbw      4 2013-02-20  review   
...                       ...    ...        ...     ...   
22951  s1dex3Z3QoqiK7V-zXUgAw      4 2013-02-19  review   
22952  PX3e3qtBx_5VC3vqq47jpQ      4 2012-09-24  review   
22953  SyXz4OwSNxfbszTHP-d2aA      5 2013-02-05  review   
22954  soH8ekCER45AjfMaM6cJ2A      4 2011-06-28  review   
22955  lHr-aMBHh29zLpuB2ShAMQ      3 2012-08-24  review   

                                              split_text  text_length  \
0      [nice, place, big, patio, now, offering, live,...          382   
1      [friendly, staff, make, sure, you, order, the,...           75   
2      [love, love, love, this, place, for, breakfast...          164   
3      [disgusting, sandwich, i, should, have, known,...         1853   
4      [always, a, fan, of, cafe, zupas, and, their, ...          459   
...                                                  ...          ...   
22951  [for, a, wine, bar, i, would, have, expected, ...          279   
22952  [this, place, has, awesome, kimchi, chigae, ka...          167   
22953  [love, this, place, i, recently, did, a, searc...          657   
22954  [excellent, inexpensive, and, hot, fresh, food...           67   
22955  [we, come, here, often, as, our, favorite, dis...          839   

                                        evaluative_words  top_words_count  
0      [nice, comedy, holiday, holiday, easily, offen...                2  
1                                       [friendly, sure]                2  
2                              [love, love, love, yummy]                4  
3      [disgusting, better, like, cutting, loyal, fan...               18  
4      [fan, friendly, amazing, great, delicious, wan...                8  
...                                                  ...              ...  
22951   [great, special, hesitation, good, trusts, sure]                4  
22952                          [awesome, ok, miss, best]                4  
22953  [love, excellent, encouraged, glad, good, inti...                7  
22954                          [excellent, fresh, worth]                3  
22955  [favorite, prickly, well, favorite, amazing, o...                9  

[22956 rows x 11 columns]

In [69]:
temp_review = df_review[['user_id', 'business_id', 'stars', 'text_length', 'top_words_count']].copy()
temp_business = df_business[['business_id', 'review_count', 'isOpen']].copy()
temp_checkin = df_checkin[['business_id', 'checkin_nums']].copy()
temp_user = df_user[['user_id']].copy()

merged_df = pd.merge(temp_review, temp_business, on = 'business_id', how = 'inner')
merged_df = pd.merge(merged_df, temp_user, on = 'user_id', how = 'inner')
merged_df = pd.merge(merged_df, temp_checkin, on = 'business_id', how = 'inner')

In [70]:
merged_df

user_id             business_id  stars  text_length  \
0     eHWbF0k5QOBLgQXhGdeHmg  8i5hB_dmf33NVbWE5SwoMQ      5           75   
1     ROJ2hXmcQc_r5ZEek1hU6w  8i5hB_dmf33NVbWE5SwoMQ      5          525   
2     T-9NA7VJmZoNhY8pTDUdHg  8i5hB_dmf33NVbWE5SwoMQ      4          156   
3     FmW2nQXyzmrz3EJzQ6v7lg  8i5hB_dmf33NVbWE5SwoMQ      3         1721   
4     1zIIIiVOZCWkI_DJ6Pz45Q  b-t3nNmRw6alKw4aSqAgPg      1          192   
...                      ...                     ...    ...          ...   
1648  WdpjLmfyOOZ8Az11WBPqkA  n8xZIU1PZz3Um0QhJnCfcw      2          357   
1649  WiEK-GiAAvqD_UYy0aa7Sg  ARsdp6eSkfPnKlPaHcv-Lg      2         1877   
1650  ushY4-xrl8uRyhrYKZabPQ  UiJajPGYV0CMI_2rV0WSRA      1          456   
1651  j-sI9L9W8tnNfdDMxAWjBg  hYooRvt98mjOun91EyKPNg      4           45   
1652  k1kbxGFR1cuVnc44pEyoxQ  dUwx303wshhFdVkc0SQgfg      1          412   

      top_words_count  review_count  isOpen  checkin_nums  
0                   2            21       1            51  
1                   8            21       1            51  
2                   5            21       1            51  
3                  12            21       1            51  
4                   3            37       1           126  
...               ...           ...     ...           ...  
1648                1             3       1            33  
1649                8             3       1            17  
1650                5             3       1             8  
1651                0             3       1            14  
1652                6             3       1             7  

[1653 rows x 8 columns]

In [71]:
# split the data
from sklearn.model_selection import train_test_split

# Assuming your DataFrame is named 'df'
train_data, test_data = train_test_split(merged_df, test_size=0.2, random_state=42)

In [73]:
X_test_scaled = scaler.transform(merged_df[['stars', 'text_length', 'top_words_count', 'review_count', 'isOpen', 'checkin_nums']])


In [74]:
predictions = model.predict(X_test_scaled)


52/52 [==============================] - 0s 2ms/step


In [77]:
print(predictions)

[[1.3642898]
 [3.7080479]
 [1.4447943]
 ...
 [3.2315583]
 [1.7014813]
 [3.1379821]]
